blocks/search.py

```diff
@@ -0,0 +1,266 @@
+        """Computes next states, glimpses, outputs, and probabilities.
+
+        Parameters
+        ----------
+        contexts : dict
+            Dictionary of contexts divided by chunks.
+        cur_vals : dict
+            Dictionary of current states, glimpses, and outputs.
+
+        Returns
+        -------
+        Dictionary of next state, glimpses, output, probabilities values
+        with names as returned by `generate_outputs`.
+
+        """
+        next_values = self.next_computer(*(contexts.values() +
+                                           [cur_vals['states'][0]]))
+        next_values = [state.reshape((1,) + state.shape)
+                       for state in next_values]
+        return self.generate_names + ['probs'], next_values
+
+    @classmethod
+    def _top_probs(cls, probs, beam_size, unique=False):
+        """Returns indexes of elements with highest probabilities.
+
+        Parameters
+        ----------
+        probs : numpy array
+            A 3d array of probabilities (length of sequence, batch,
+            readout_dim)
+        beam_size : int
+            Beam size, number of top probs to return
+
+        Returns
+        -------
+        Tuple of (indexes, top probabilities)
+
+        """
+        flatten = probs.flatten()
+        if unique:
+            args = numpy.unique(
+                numpy.argpartition(-flatten, beam_size))[:beam_size]
+        else:
+            args = numpy.argpartition(-flatten, beam_size)[:beam_size]
+        args = args[numpy.argsort(-flatten[args])]
+        if unique:
+            # append best if needed
+            if args.shape[0] < beam_size:
+                args = numpy.append(
+                    args,
+                    numpy.tile(args[0], beam_size - args.shape[0]))
+        # convert args back
+        indexes = numpy.unravel_index(args, probs.shape[1:])
+        return indexes, probs[0][indexes]
+
+    @staticmethod
+    def _rearrange(outputs, indexes):
+        new_outputs = outputs[:, indexes.T.flatten()]
+        return new_outputs.copy()
+
+    def search(self, inputs_val_dict, eol_symbol=-1, max_length=512):
+        """Performs beam search.
+
+        Parameters
+        ----------
+        inputs_val_dict : dict
+            Dictionary of input values {name: value}. Input values may
+            be in a batch of size greater than 1, in this case, several
+            beam searches will be performed in parallel.
+        eol_symbol : int
+            End of sequence symbol, the search stops when the
+            symbol is generated.
+        max_length : int
+            Maximum sequence length, the search stops when it
+            is reached.
+
+        Returns
+        -------
+        Most probable sequences, corresponding probabilities and costs.
+
+        """
+        if not self.compiled:
+            self.compile()
+        # Inputs repeated beam_size times
+        aux_inputs = OrderedDict(
```

Comment:
I know I am terrible picky, but "auxiliary" more confuses here than explain what happens. Together with your nice informative comment simply "inputs" will do the job.